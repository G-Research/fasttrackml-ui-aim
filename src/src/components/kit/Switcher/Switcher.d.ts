import React, { ReactElement } from 'react';

export interface ISwitcherProps {
  leftLabel?: SwitcherLabel;
  rightLabel?: SwitcherLabel;
  checked: boolean | undefined;
  color?: 'primary' | 'secondary';
  size?: 'small' | 'medium' | 'large';
  variant?: 'contained' | 'outlined';
  name?: string;
  onChange: (e: React.ChangeEvent, checked: boolean) => void;
}

export type SwitcherLabel =
  | string
  | number
  | React.FunctionComponent<React.ReactElement>
  | ReactElement;
