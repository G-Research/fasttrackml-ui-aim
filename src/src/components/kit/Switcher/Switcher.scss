@use 'src/styles/abstracts' as *;

.Switcher {
  display: inline-block;
  cursor: pointer;
  border-radius: 0.625rem;
  color: #fff;
  position: relative;
  user-select: none;
  border: none;

  &__circle {
    border-radius: 0.625rem;
    background-color: #fff;
    transition: all 0.1s linear;
    position: absolute;
    left: 2px;
    top: $space-xxxxs;
  }

  &__leftLabel,
  &__rightLabel {
    transition: all 0.18s ease-out;
    position: absolute;
    top: 4px;
    line-height: 8px;
  }

  &__leftLabel {
    left: 4px;
    transform: scale(0);
  }

  &__rightLabel {
    transform: scale(1);
    right: 4px;
  }

  &__checked {
    .Switcher__leftLabel {
      transform: scale(1);
    }

    .Switcher__rightLabel {
      transform: scale(0);
    }
  }

  &__primary {
    background-color: $pico-50;

    &.Switcher__checked {
      background-color: $primary-color;
    }
  }

  &__secondary {
    background-color: $pico-50;

    &.Switcher__checked {
      background-color: $pico;
    }
  }

  &__small {
    height: 0.8125rem;
    width: 1.5rem;
    font-size: 0.5rem;
    .Switcher__circle {
      width: 0.5625rem;
      border-radius: 0.5rem;
      height: 0.5625rem;
      line-height: 0.5625rem;
      top: 2px;
      &__checked {
        left: calc(100% - 11px);
      }
    }
    &:hover {
      .Switcher__circle {
        width: 0.675rem;
        &__checked {
          left: calc(100% - 13px);
        }
      }
    }
    .Switcher__leftLabel,
    .Switcher__rightLabel {
      top: 3px;
      font-size: 7px;
      line-height: 7px;
      display: flex;
      align-items: center;
      justify-content: center;
    }
  }

  &__medium {
    height: 1rem;
    width: 2rem;
    line-height: 8px;

    .Switcher__circle {
      font-size: 0.75rem;
      width: 0.75rem;
      height: 0.75rem;
      &__checked {
        left: calc(100% - 14px);
      }
    }
    &:hover {
      .Switcher__circle {
        width: 0.9375rem;
        &__checked {
          left: calc(100% - 17px);
        }
      }
    }
  }

  &__large {
    height: 1.25rem;
    width: 2.5rem;
    font-size: 10px;
    line-height: 10px;

    .Switcher__circle {
      width: 1rem;
      height: 1rem;
      &__checked {
       left: calc(100% - 18px);
      }
    }
    &:hover {
      .Switcher__circle {
        width: 1.25rem;
        &__checked {
          left: calc(100% - 22px);
        }
      }
    }

    .Switcher__leftLabel,
    .Switcher__rightLabel {
      font-size: 10px;
      line-height: 8px;
      display: flex;
      align-items: center;
      justify-content: center;
      top: 5px;
    }
  }

  &__outlined {
    border: 0.0625rem solid $pico-50;
    background-color: #fff;
    color: $pico-50;

    &.Switcher__primary {
      &.Switcher__small {
        .Switcher__rightLabel {
          top: 2px;
          right: 2px;
        }

        .Switcher__leftLabel {
          top: 2px;
          left: 2px;
        }
      }

      .Switcher__circle {
        color: $pico-50;
        background-color: $pico-50;
        top: 1px;
        &__checked {
          color: $primary-color;
          background-color: $primary-color;
        }
      }

      &.Switcher__checked {
        border-color: $primary-color;
        background-color: #fff;
        color: $primary-color;
      }
    }
  }
}
