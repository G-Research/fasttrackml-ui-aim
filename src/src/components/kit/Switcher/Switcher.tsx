import React from 'react';

import ErrorBoundary from 'components/ErrorBoundary/ErrorBoundary';

import { ISwitcherProps } from './Switcher.d';

import './Switcher.scss';

function Switcher({
  onChange,
  checked,
  color = 'primary',
  leftLabel,
  rightLabel,
  size = 'medium',
  variant = 'contained',
  name = 'switcher',
}: ISwitcherProps) {
  const [checkedValue, setCheckedValue] = React.useState<boolean | undefined>(
    checked,
  );

  function handleClick(e: React.ChangeEvent<any>) {
    setCheckedValue(!checkedValue);
    onChange(e, !checkedValue);
  }

  React.useEffect(() => {
    setCheckedValue((prevState) => {
      return checked !== prevState ? checked : prevState;
    });
  }, [checked]);

  return (
    <ErrorBoundary>
      <button
        data-name={name}
        data-testid='switcher'
        className={`Switcher ${`Switcher__${variant}`} ${`Switcher__${color}`} ${`Switcher__${size} ${
          checkedValue ? 'Switcher__checked' : ''
        }`}`}
        onClick={handleClick}
      >
        {leftLabel && <span className='Switcher__leftLabel'>{leftLabel}</span>}
        <i
          className={`Switcher__circle ${
            checkedValue ? 'Switcher__circle__checked' : ''
          }`}
        />
        {rightLabel && (
          <span className='Switcher__rightLabel'>{rightLabel}</span>
        )}
      </button>
    </ErrorBoundary>
  );
}

export default React.memo(Switcher);
