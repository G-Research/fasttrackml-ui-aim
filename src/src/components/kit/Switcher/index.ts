import Switcher from './Switcher';

export * from './Switcher.d';
export default Switcher;
