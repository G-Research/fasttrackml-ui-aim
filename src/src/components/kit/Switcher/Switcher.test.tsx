import { render, fireEvent } from '@testing-library/react';

import Switcher from './Switcher';

const mockClick = jest.fn();

describe('<Switcher /> -', () => {
  test('Renders Correctly', () => {
    const { asFragment } = render(
      <Switcher
        onChange={mockClick}
        checked
        rightLabel='right label'
        leftLabel='left label'
      />,
    );

    expect(asFragment()).toMatchSnapshot();
  });

  test('Renders width labels correctly', () => {
    const { getByText } = render(
      <Switcher
        onChange={mockClick}
        rightLabel='right label'
        leftLabel='left label'
        checked
      />,
    );

    expect(getByText('right label', { exact: true }));
    expect(getByText('left label', { exact: true }));
  });

  test('onchange event works properly', () => {
    const { getByTestId } = render(<Switcher onChange={mockClick} checked />);

    const switcher = getByTestId('switcher');

    fireEvent.click(switcher);
    expect(mockClick).toHaveBeenCalledTimes(1);
    expect(switcher.querySelector('i')).not.toHaveClass(
      'Switcher__circle__checked',
    );

    fireEvent.click(switcher);
    expect(mockClick).toHaveBeenCalledTimes(2);
    expect(switcher.querySelector('i')).toHaveClass(
      'Switcher__circle__checked',
    );
  });
});
