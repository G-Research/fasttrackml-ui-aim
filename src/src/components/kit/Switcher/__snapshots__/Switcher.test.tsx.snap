// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Switcher /> - Renders Correctly 1`] = `
<DocumentFragment>
  <button
    class="Switcher Switcher__contained Switcher__primary Switcher__medium Switcher__checked"
    data-name="switcher"
    data-testid="switcher"
  >
    <span
      class="Switcher__leftLabel"
    >
      left label
    </span>
    <i
      class="Switcher__circle Switcher__circle__checked"
    />
    <span
      class="Switcher__rightLabel"
    >
      right label
    </span>
  </button>
</DocumentFragment>
`;
